### DISCUSSION TITLE
Ideas for IPv6 implementation in Orjail

### DISCUSSION THREAD

**Discussion by phantomcraft** [orjail/orjail#84] [Ideas]
There is no "Discussion" section in this repository, so I'm opening an issue. It's not that hard to implement IPv6 support in Orjail.

========================================================================

IPv6 address for local addressing (ULAs) can be generated in this way (considering finding a valid address in the system):
```

VETH_NET_ADDR6="fd$(printf "%x\n" $((RANDOM%238 + 16))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1)))"
  if [ "inet ${VETH_NET_ADDR6}::1" = "$(ip address show | grep ${VETH_NET_ADDR6}::1 | cut -d "/" -f 1)" ]; then
    while [ "inet ${VETH_NET_ADDR6}::1" = "$(ip address show | grep ${VETH_NET_ADDR6}::1 | cut -d "/" -f 1)" ]; do
      VETH_NET_ADDR6="fd$(printf "%x\n" $((RANDOM%238 + 16))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1)))"
    done
  fi
VETH_ADDR6="${VETH_NET_ADDR6}::1"
PEER_ADDR6="${VETH_NET_ADDR6}:$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1)))
NETMASK6="64"
```

========================================================================

An example (adjust the variables and names and all the stuff in Orjail mode):

```
VETH_NET_ADDR6="fd$(printf "%x\n" $((RANDOM%238 + 16))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1)))"
  if [ "inet ${VETH_NET_ADDR6}::1" = "$(ip address show | grep ${VETH_NET_ADDR6}::1 | cut -d "/" -f 1)" ]; then
    while [ "inet ${VETH_NET_ADDR6}::1" = "$(ip address show | grep ${VETH_NET_ADDR6}::1 | cut -d "/" -f 1)" ]; do
      VETH_NET_ADDR6="fd$(printf "%x\n" $((RANDOM%238 + 16))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1)))"
    done
  fi
VETH_ADDR6="${VETH_NET_ADDR6}::1"
PEER_ADDR6="${VETH_NET_ADDR6}:$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1))):$(printf "%x\n" $((RANDOM%65533 + 1)))"
NETMASK6="64"
TRANSPORT=9040
DNSPORT=5354

[...]

ip6tables -t nat -I POSTROUTING 1 \! -o ${VETH} -s ${VETH_NET_ADDR6}/{NETMASK6} -j RETURN
ip6tables -t nat -I PREROUTING 1 \! -i ${VETH} -d ${VETH_NET_ADDR6}/{NETMASK6} -j RETURN
ip6tables -t nat -A PREROUTING -i ${VETH} -j RETURN
ip6tables -t nat -A PREROUTING -i ${VETH} -p udp -d ${VETH_ADDR6} --dport 53 -j DNAT --to-destination "[${VETH_ADDR6}]":${DNSPORT}
ip6tables -t nat -A PREROUTING -i ${VETH} -p tcp --syn -j DNAT --to-destination "[${VETH_ADDR6}]":${TRANSPORT}
ip6tables -A INPUT ! -i ${VETH} -s ${VETH_NET_ADDR6}/{NETMASK6} -j DROP
ip6tables -A INPUT ! -i ${VETH} -d ${VETH_NET_ADDR6}/{NETMASK6} -j DROP
ip6tables -A INPUT -i ${VETH} -p ipv6-icmp -m ipv6-icmp --icmpv6-type 135 -s ${PEER_ADDR6} -d ff02::1:ff00:1 -j ACCEPT
ip6tables -A INPUT -i ${VETH} -p ipv6-icmp -m ipv6-icmp --icmpv6-type 136 -s ${PEER_ADDR6} -d ${VETH_ADDR6} -j ACCEPT
ip6tables -I INPUT -i ${VETH} -d ${VETH_ADDR6} -p udp --dport ${DNSPORT} -j ACCEPT
ip6tables -I INPUT -i ${VETH} -d ${VETH_ADDR6} -p udp --dport ${TRANSPORT} -j ACCEPT
ip6tables -A INPUT -i ${VETH} -j DROP
ip6tables -A OUTPUT -m state -o ${VETH} --state ESTABLISHED,RELATED -j ACCEPT
ip6tables -I FORWARD -i ${VETH} -j DROP
ip6tables -I FORWARD -o ${VETH} -j DROP

[...]

sysctl -w -q "net.ipv6.conf.${VETH}".forwarding=0

[...]

echo "VirtualAddrNetworkIPv6 [${PEER_ADDR6}]/48
DNSPort ${VETH_ADDR6}:${DNSPORT} IPv6Traffic PreferIPv6
TransPort ${VETH_ADDR6}:${TRANSPORT}
ClientUseIPv6 1
ClientPreferIPv6ORPort 1" >> ${TORCONFIGFILE}
```

/\ The rest are in the same way as IPv4, as adding addresses to Veth, configuring the internal route inside network namespace. And always consider the ff02::1:ff00:1 address, it will not work without it.

VirtualAddrNetworkIPv6 is required to have the peer address and the "48" suffix.

DNSPort *MUST* have IPv6Traffic and PreferIPv6 options for (probably) finding the IPv6 enabled exit node, and Tor builds a secondary circuit just for DNS resolving.

**ClientUseIPv6** and **ClientPreferIPv6ORPort** are required.

========================================================================

There are some problems in Tor IPv6 implementation:

=> DNS port not always return AAAA addresses.
=> Tor doesn't have a automatic way of connecting to IPv6 enabled exit relays, it chooses the nodes in a circuit randomly, there is a great chance of picking up a IPv4 only circuit.
=> If the client doesn't have IPv6 connection, the chance of picking up a IPv6 enabled tunnel is minimal.

=========================================================================

@lesion

Any questions, post here and soon I will respond.

**Comment by lesion** [comment]
Migrated to "discussion". I had no idea that github has implemented this discussion feature.
ipv6 support is difficult to test for me as I don't have ipv6 support anywhere. I'm using the only vps with an ipv6 address available and I cannot make it works (I'm stuck at resolving addresses, using for example `curl -6 https://ipv6.icanhazip.com` ).
I'll keep trying. PR to `ipv6` branch welcome of course. Thanks!

**Comment by phantomcraft** [comment]
@lesion 

No problem if you don't have IPv6 connectivity, just set up a tunnel broker:

https://duckduckgo.com/?q=tunnel+broker+ipv6+how+to&ia=web

https://tunnelbroker.net

For Hurricane Electric tunnel is very easy to you to configure if you use a router and have 192.168.1.x as local address:
```

modprobe ipv6
ip tunnel add he-ipv6 mode sit remote <remote_tunnel_ip> local 192.168.1.x ttl 255
ip link set he-ipv6 up
ip addr add <ipv6_address_of_tunnel>/64 dev he-ipv6
ip route add ::/0 dev he-ipv6
ip -f inet6 addr
```

![Screenshot_2021-09-21 Tunnel Details](https://user-images.githubusercontent.com/39815480/134235836-471470ec-3d7b-44f5-8a3a-8afb89506afe.png)

**Comment by phantomcraft** [comment]
@lesion

> Migrated to "discussion". I had no idea that github has implemented this discussion feature. ipv6 support is difficult to test for me as I don't have ipv6 support anywhere. I'm using the only vps with an ipv6 address available and I cannot make it works (I'm stuck at resolving addresses, using for example `curl -6 https://ipv6.icanhazip.com` ). I'll keep trying. PR to `ipv6` branch welcome of course. Thanks!

It's not that hard to have IPv6 connectivity in Tor if the user doesn't have IPv6.

All that should be done is to get a list of all Tor exit nodes with IPv6 and add to torrc, try to run this shell combo to get a list of them (comma separated fingerprints):

`grep -B3 '^s\ Exit\ ' /var/lib/tor/cached-microdesc-consensus | awk '/^a \[/{print a $0}{a=$0 ORS}' | grep '^r\ ' | cut -d " " -f 3 | while read line; do echo "${line}" | base64 -d 2>/dev/null | xxd -p; done | while read line; do echo -en \$"${line}", ; done; echo ''`

In Orjail Tor must be ran 2 times to enable IPv6, one for getting a `consensus list` and add the specific exit relays to torrc, and the next to connect Orjail.

You should add this options to get the consensus over IPv4:

```
Log info file /tmp/orjail-$NAME/log
ClientUseIPv6 0
ClientUseIPv4 1
ClientPreferIPv6ORPort 0
```

Use a shell loop like this to know the exact moment that Tor consensus is fully loaded:

```
while [ ! $(cat /tmp/orjail-$NAME | grep 'Bootstrapped 100% (done): Done') ]; do
  sleep 0.5
done
```

Add a shell line to kill Tor, and add these lines to torrc:

```
ClientUseIPv6 1
ClientUseIPv4 1
ClientPreferIPv6ORPort 1
StrictNodes 1
```

And add this line to the script (it will add all the exit relays with IPv6):

`echo "ExitNodes $(grep -B3 '^s\ Exit\ ' /tmp/orjail-$NAME/cached-microdesc-consensus | awk '/^a \[/{print a $0}{a=$0 ORS}' | grep '^r\ ' | cut -d " " -f 3 | while read line; do echo "${line}" | base64 -d 2>/dev/null | xxd -p; done | while read line; do echo -en \$"${line}",; done)" >> $TORCONFIGFILE`

The result will be Tor with 100% IPv6 connectivity even if the user is IPv4 only.

The only disadvantage is that this trick would require **vim** (in Debian, **xxd** package). **grep** and **awk** wouldn't be a problem as it's available in almost all Linux distributions.
